### PR TITLE
feat: add API option to disable Alertmanager deployment

### PR DESCRIPTION
**Pull request by tremes**
This adds a new configuration option for the MonitoringStack CRD that
allows users to disable Alertmanager deployment.

This is a first draft. I would like to add:
- removing of other Alermanager's resources (service, role, etc.)
- some tests

### CODE REVIEW

File: pkg/controllers/monitoring/monitoring-stack/controller.go
Comment (tremes) [inline pkg/controllers/monitoring/monitoring-stack/controller.go:217]: Lower case and typo....note to myself :)

**Review by simonpasquier — COMMENTED** [review]
I think that we could have 2 types of reconciler:
* patchReconciler which patches an existing resource to the desired state (e.g. defaultReconciler as it exists today).
* deleteReconciler which ensures that a given resource doesn't exist in the cluster.

`stackComponentReconcilers()` would know which reconciler to use for which resource.

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (sthaha) [inline pkg/apis/monitoring/v1alpha1/types.go:78]: I think it is better to have an `AlertManagerConfig` and have `Disabled` under it. So that the spec reads
```
spec:
  alertmanager:
    disabled: true
```

Comment (tremes) [reply pkg/apis/monitoring/v1alpha1/types.go:78]: I've thought about it and I think we can do it, but at the same time it seems like the only config option for the Alertmanager (but one never knows....).

Comment (sthaha) [reply pkg/apis/monitoring/v1alpha1/types.go:78]: Yes, the idea is to namespace so that additional configs do not require changes to existing api.

File: test/e2e/monitoring_stack_controller_test.go
Comment (simonpasquier) [inline test/e2e/monitoring_stack_controller_test.go:548]: (nit) to avoid grow the list of arguments indefinitely, I'd recommend using something like the functional options pattern.

```golang
func newMonitoringStack(t *testing.T, name string, options ...func(*stack.MonitoringStack)) *stack.MonitoringStack {
    ms := &stack.MonitoringStack{
    ...
    }
    for _, opt := range opts {
        opt(ms)
    }
    ...
    return ms
}
```

Then use it like

```
newMonitoringStack(t, name,
    func(ms *stack.MonitoringStack) {
        ms.Spec.AlertmanagerConfig.DisableAlertmanager = true
    },
)
```

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/types.go:178]: Do we need to repeat Alertmanager?

```suggestion
	Disabled bool `json:"disabled,omitempty"`
```

**Review by simonpasquier — COMMENTED** [review]
I think that the operator should disable [Alertmanager configuration](https://github.com/rhobs/observability-operator/blob/93a144b0bea739752dfdfd7b4c6f59944ac9e6bd/pkg/controllers/monitoring/monitoring-stack/components.go#L191-L201) in Prometheus too.

File: test/e2e/framework/assertions.go
Comment (sthaha) [inline test/e2e/framework/assertions.go:283]:
```suggestion
func (f *Framework) AssertAlertmanagerAbsent(t *testing.T, name, namespace string) {
```

File: test/e2e/framework/assertions.go
Comment (sthaha) [inline test/e2e/framework/assertions.go:297]: i
```suggestion
		t.Fatal(fmt.Errorf("alertmanager %s/%s is present when expected to be absent", namespace, name))
```

File: test/e2e/monitoring_stack_controller_test.go
Comment (sthaha) [reply test/e2e/monitoring_stack_controller_test.go:548]: calling `newMonitoringStack(t, "foobar", false)` is confusing to read. [The infamous boolean trap!](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap)

while `newMontoringStack(t, "foobar", withoutAlertManager)` is clearer 
where `func withoutAlertManager(ms *stack.MonitoringStack) { ms.Spec.AlertmanagerConfig.Disabled = true }`

**Comment by tremes** [comment]
> I think that the operator should disable [Alertmanager configuration](https://github.com/rhobs/observability-operator/blob/93a144b0bea739752dfdfd7b4c6f59944ac9e6bd/pkg/controllers/monitoring/monitoring-stack/components.go#L191-L201) in Prometheus too.

@simonpasquier  I guess it means to remove the spec and the endpoint right (there's no extra switch right, when looking at https://pkg.go.dev/github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1#PrometheusSpec)?

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (tremes) [reply pkg/apis/monitoring/v1alpha1/types.go:178]: we don't :)

**Comment by simonpasquier** [comment]
@tremes correct you need something like

```golang
prometheus := &monv1.Prometheus{
...
}

if !ms.Spec.AlertmanagerConfig.DisableAlertmanager {
    prometheus.Spec.Alerting = &monv1.AlertingSpec{
        ....
    }
} 
```

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/types.go:176]:
```suggestion
	// Disables the deployment of Alertmanager. 
	// Note that in this case, Prometheus doesn't send alerts. Disabling Alertmanager is mainly useful when using the monitoring stack in a collect-and-forward mode.
```

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (jan--f) [inline pkg/controllers/monitoring/monitoring-stack/components.go:95]: Another option would be to have the reconcile function inspect `ms.Spec.AlertmanagerConfig.Disabled` similarly to what `prometheusPDBReconciler` is doing.
This would avoid treating the alertmanager resources differently in this location.
But that's nothing that prevents this from merging...

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/types.go:78]: This avoids checking for nil values.

```suggestion
	AlertmanagerConfig AlertmanagerConfig `json:"alertmanagerConfig,omitempty"`
```

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (simonpasquier) [reply pkg/controllers/monitoring/monitoring-stack/components.go]: Or instead of `prometheusPDBReconciler`, the controller could decide to use `defaultReconciler` (replicas > 1) or `removeReconciler`.

Comment (tremes) [reply pkg/controllers/monitoring/monitoring-stack/components.go]: Hmm so the suggestions are either have multiple functions and keep `stackComponentReconcilers` simple or have two functions (`defaultReconciler` and `removeReconciler` + possibly more in the futture) and have the logic in the `stackComponentReconcilers` right? 

I think both are possible...

Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: Just to offer a third version:
We could have a generic reconciler function that takes an additional argument: `removeReconciler(component client.Object, ms *stack.MonitoringStack, deleteCondition bool) reconcileFunction`.

I'm not the biggest fan of repeating blocks of
```
if condition1 {
		reconcileFunctions = append(reconcileFunctions, removeReconciler(obj1, ms))
} else {
	reconcileFunctions = append(reconcileFunctions, defaultReconciler(obj1, ms))
}
if condition2 {
		reconcileFunctions = append(reconcileFunctions, removeReconciler(obj2, ms))
} else {
	reconcileFunctions = append(reconcileFunctions, defaultReconciler(obj2, ms))
}
...
```
Is there and advantage of the above I'm not seeing?

Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: also to reiterate: I don't think this discussion needs to hold up this PR. I have a branch to put the reconcile logic in a module and use it in the thanos querier controller too.

Comment (tremes) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: > Is there and advantage of the above I'm not seeing?

I don't think so. If you have it prepared in your branch then I would leave it.

Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: See #185 for what I had in mind. I'd very much appreciate feedback.

With that in mind, I'm happy to move this forward and discuss the reconcile bit on the other PR.

Comment (sthaha) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: I feel the code reads cleaner if the decision making part moves to a separate function that returns the right reconciler to use. 

e.g. the code would read as 


```go
reconcilers =  []reconcileFunction {
   defaultReconciler(...),
   pdbReconciler(ms), 
}

reconcilers = append(reconcilers, ...alertmanagerReconcilers(ms))
return reconcilers
```
where `pdbReconciler` will return the `remove|default` reconciler
and `alertmanagerReconcilers` will return the appropiate reconcilers

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (sthaha) [inline pkg/controllers/monitoring/monitoring-stack/components.go:45]:
```suggestion
func deleteReconciler(component client.Object, ms *stack.MonitoringStack) reconcileFunction {
```

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (sthaha) [inline pkg/apis/monitoring/v1alpha1/types.go:178]: can we also set `default: false`?

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (sthaha) [inline pkg/apis/monitoring/v1alpha1/types.go:78]:
```suggestion
	// +kubebuilder:default={disabled: false}
	AlertmanagerConfig *AlertmanagerConfig `json:"alertmanagerConfig,omitempty"`
```

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (sthaha) [reply pkg/controllers/monitoring/monitoring-stack/components.go:45]: To match `client.Delete`

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (tremes) [reply pkg/apis/monitoring/v1alpha1/types.go:178]: Is that required? It's the zero value.

Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/types.go:178]: > Is that required? It's the zero value.

agreed, I don't feel we need to specify a default.

Comment (sthaha) [reply pkg/apis/monitoring/v1alpha1/types.go:178]: We need to specify default so that it appears when `oc get -o yaml` an already applied resource where user has relied on the defaults. see: replicas e.g.

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (tremes) [reply pkg/controllers/monitoring/monitoring-stack/components.go:100]: So do you want me to update it? It seems unnecessary given the Jan's proposal (in different PR).

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (tremes) [reply pkg/apis/monitoring/v1alpha1/types.go:79]: OK

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (tremes) [reply pkg/controllers/monitoring/monitoring-stack/components.go]: OK

File: pkg/apis/monitoring/v1alpha1/types.go
Comment (tremes) [reply pkg/apis/monitoring/v1alpha1/types.go:180]: OK

**Comment by simonpasquier** [comment]
I'd be fine with merging this one and discussing the follow-up options in #185.

**Review by sthaha — DISMISSED** [review]

**Review by sthaha — CHANGES_REQUESTED** [review]
@tremes I had approved it and was going to merge when I realised that the commit-lint fails :(

Could you please fix and resubmit?

**Comment by JoaoBraveCoding** [comment]
Merging it as per @sthaha comment

**Review by JoaoBraveCoding — APPROVED** [review]

File: pkg/controllers/monitoring/monitoring-stack/components.go
Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/components.go]: I'm ok with leaving this as is.

**Review by jan--f — APPROVED** [review]

**Review by sthaha — APPROVED** [review]

**Pull request merged**
